### ISSUE TITLE
Any assocation with an extension block raises the deprecation warning about instance-dependence

### ISSUE THREAD

**Issue by jaroslawr** [rails/rails#18467]
With an association defined like this:

```ruby
class Parent < ActiveRecord::Base
  has_many :children, ->() { order('id DESC') }, dependent: :destroy do
    def xyzyyx
       ...
    end
    ...
  end
end
```

and a query like this:

```ruby
Parent.includes(:children).to_a
```

you get the deprecation warning about instance-depedence:

```
DEPRECATION WARNING: The association scope 'children' is instance dependent (the scope block takes an argument). Preloading happens before the individual instances are created. This means that there is no instance being passed to the association scope. This will most likely result in broken or incorrect behavior. Joining, Preloading and eager loading of these associations is deprecated and will be removed in the future. (called from test_preloading_of_associations at preloading_test.rb:25)
```

which does not seem to be sensible in this case.

Runnable test case:

```ruby
gem 'activerecord', '4.2.0'
require 'active_record'
require 'minitest/autorun'

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.default_timezone = :local

ActiveRecord::Schema.define do
  create_table(:parents)
  create_table(:children) { |t| t.integer :parent_id }
end

class Parent < ActiveRecord::Base
  has_many :children, ->() { order('id DESC') }, dependent: :destroy do
  end
end

class Child < ActiveRecord::Base
end

class Bug < Minitest::Test
  def test_preloading_of_associations
    Parent.create!
    Parent.create!
    Parent.includes(:children).to_a
  end
end
```

The culprit seems to be in CollectionAssociation in ActiveRecord::Associations::Builder, the initialize method says:

```ruby
def initialize(model, name, scope, options)
  super
  @mod = nil
  if block_given?
    @mod = Module.new(&Proc.new)
    @scope = wrap_scope @scope, @mod
  end
end
```
And wrap_scope is:

```ruby
def wrap_scope(scope, mod)
  if scope
    proc { |owner| instance_exec(owner, &scope).extending(mod) }
  else
    proc { extending(mod) }
  end
end
```

With this you end up with scope.arity > 0 even though there is no user-given scope (or the user-given scope actually is of arity 0).

@senny what do you think about this?

**Comment by senny** [comment]
@jaroslawr thank you for reporting, I'll take a look.

**Comment by senny** [comment]
@jaroslawr I just pushed a patch to the `4-2-stable` branch. Would be great to hear wether this solves your issue.

Will be released with the next patch version of `4.2.`.
